CDS-46002-Spot Elastigroup deployment (#1702)
harness/developer-hub: docs/continuous-delivery/deploy-srv-diff-platforms/aws/spot/spot-deployment.md

---
title: Spot Elastigroup deployments
description: This topic walks you through deploying Spot Elastigroup using Harness.
sidebar_position: 1
---

:::info

Currently, Spot Elastigroup support is behind the feature flag `SPOT_ELASTIGROUP_NG`. Contact [Harness Support](mailto:support@harness.io) to enable the feature.

:::

> This topic describes how to configure and execute Amazon Machine Image (AMI) deployments through the Spot Elastigroup management platform. 

Reviewer: ```suggestion
This topic describes how to deploy AWS ASG deployments using Harness and the Spot Elastigroup management platform. 
```

11. In **Test Region**, select an AWS region to test the credentials for this connector. 
12. In **Delegates Setup**, select **Only use Delegate with all of the following tags** and enter the name of the delegate created in [connect to a Spot cloud provider (step 7)](#connect-to-a-spot-cloud-provider).
13. Select **Save and Continue**
14.  After the connection test succeeds, select **Continue**.
15. In **Artifact Details**, enter the following details:
    1.  In **Artifact Source Identifier**, enter a name for the artifact.
    2.  In **Region**, select the region where the artifact is located. 
    3.  (Optional) In **AMI Tags**, add the tags used by your AMI. 
    4.  (Optional) In **AMI Filters**, add the AMI ID filters to locate the AMI resource.
    5.  In **Version Details**, select **Value** to enter the version number of the package you want to use. You can also select **Regex**, and then enter a tag regex to filter the package version.
    
    ![](./static/spot-artifacts-ami.png)
16. Select **Submit**.
17. **Save** the service configuration.

> <details>

Reviewer: I would put the Pipeline Studio steps in one tab and this YAML example in a second tab.

<summary>Sample service YAML</summary>

Here's a sample Spot service YAML: 

```
service:
  name: spot-svc
  identifier: spotsvc
  serviceDefinition:
    type: Elastigroup
    spec:
      artifacts:
        primary:
          primaryArtifactRef: <+input>
          sources:
            - identifier: spot_ami
              spec:
                connectorRef: org.AWSConnectorForAutomationTest
                region: us-east-1
                version: <+input>.allowedValues(asgAmi-Orderv2,asgAmi-Order)
              type: AmazonMachineImage
  gitOpsEnabled: false
```
</details>

## Add a Spot environment

Define the environment where you will deploy your application.

> 1. In **Environment >** **Specify Environment**, select **New Environment**.  

Reviewer: ```suggestion
1. In **Environment**, select **New Environment**.  
```

2. Enter the name **Spot**, and the select **Production** or **Pre-Production**.
4. Select **Save**.
5. In **Specify Infrastructure**, select **New Infrastructure**.
6. Enter a name, and then verify that the selected deployment type is **Spot Elastigroup**.
7. Select the [Spot connector](#connect-to-a-spot-cloud-provider) you created earlier.

> 8. In **Elastigroup Configuration**, select an Elastigroup configuration source file from the [Harness file store](/docs/continuous-delivery/x-platform-cd-features/services/add-inline-manifests-using-file-store/) or create a new file.

Reviewer: Do we support YAML and JSON formats? Might need to check with the Eng.

             "t3a.small",
             "t2.medium",
             "t3.medium",
             "t3a.medium"
           ]
         },
         "availabilityZones": [
           {
             "name": "us-east-1a",
             "subnetIds": [
               "subnet-1f703d78"
             ]
           },
           {
             "name": "us-east-1b",
             "subnetIds": [
               "subnet-01bdf52f"
             ]
           },
           {
             "name": "us-east-1c",
             "subnetIds": [
               "subnet-33eaf779"
             ]
           },
           {
             "name": "us-east-1d",
             "subnetIds": [
               "subnet-c1ce809d"
             ]
           },
           {
             "name": "us-east-1e",
             "subnetIds": [
               "subnet-7427b64a"
             ]
           },
           {
             "name": "us-east-1f",
             "subnetIds": [
               "subnet-11efe81e"
             ]
           }
         ],
         "product": "Linux/UNIX",
         "launchSpecification": {
           "loadBalancersConfig": {
             "loadBalancers": [
               {
                 "name": "asg-tg6",

>                  "arn": "arn:aws:elasticloadbalancing:us-east-1:839162981415:targetgroup/asg-tg6/3bcd13e204785948",

Reviewer: ```suggestion
                 "arn": "arn:aws:elasticloadbalancing:us-east-1:1234567890:targetgroup/asg-tg6/3bcd13e204785948",
```

Reviewer: try not to use actual AWS account numbers.

         }
       },
       "scaling": {},
       "scheduling": {},
       "thirdPartiesIntegration": {}
    }
   }
   ```

   </details>
9. Select **Save**.

<details>
<summary>Sample infrastructure definition YAML</summary>

Here's a sample infrastructure definition YAML: 
```
infrastructureDefinition:
  name: spot-infra
  identifier: spotinfra
  description: ""
  tags: {}
  orgIdentifier: Ng_Pipelines_K8s_Organisations
  projectIdentifier: DoNotDelete_IvanBalan
  environmentRef: Spot
  deploymentType: Elastigroup
  type: Elastigroup
  spec:
    connectorRef: spotinstconn
    configuration:
      store:
        type: Harness
        spec:
          files:
            - /spot/spot-cfg
  allowSimultaneousDeployments: false
```

</details>

## Spot Elastigroup execution strategies

Now you can select a [deployment strategy](/docs/continuous-delivery/manage-deployments/deployment-concepts.md) for this stage of the pipeline.

:::info
Spot Elastigroup deployments support only Basic, Canary, and Blue Green execution types.
:::

```mdx-code-block
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
```
```mdx-code-block
<Tabs>
<TabItem value="Basic" label="Basic">
```

> Spot Elastigroups perform the functions that Auto Scaling Groups perform in standard AMI deployments. By default, Harness Spot Elastigroup Basic deployment pre-configures the following steps:

Reviewer: ```suggestion
Spot Elastigroups perform the functions that Auto Scaling Groups perform in standard AMI deployments. By default, a Harness Spot Elastigroup Basic deployment pre-configures the following steps:
```

* Elastigroup Setup - Specify how many instances to launch, and their steady state timeout.
* Elastigroup Deploy - Specify how many instances to deploy, as a number or percentage of the Elastigroup parameters you've set up.

> ## Create a Basic deployment

Reviewer: I would include a YAML pipeline too, if you have one. Same for the other strategies.

1. In Execution Strategies, select **Basic**, then select **Use Strategy**.
2. The basic execution steps are added. 
   
   ![](./static/spot-basic.png)

3. Select the **Elastigroup Setup** step to define **Step Parameters**.
    * **Name** - Enter the deployment step name.
    * **Timeout** - Set how long you want the Harness delegate to wait for the Spot cloud to respond to API requests before timeout.

>     * **App Name** - Enter a name for the Elastigroup app name that Harness will create. For example, `<+project.identifier>_<+service.identifier>_<+env.identifier>_basic2`. You can select a fixed value, runtime input, or expression for this field. 

Reviewer: ```suggestion
    * **App Name** - Enter a name for the Elastigroup app name that Harness will create. For example, you can use Harness expressions, such as `<+project.identifier>_<+service.identifier>_<+env.identifier>_myasg`. You can select a fixed value, runtime input, or expression for this field. 
```

    * In **Instances**, select: 
       * **Same as already running instances** to replicate the already running instances.
       * **Fixed** and enter **Min Instances**, **Max Instances**, and **Desired Instances**.
4. Select the **Elastigroup Deploy** step to define **Step Parameters**.
    * **Name** - Enter the deployment step name.
    * **Timeout** - Set how long you want the Harness delegate to wait for the Spot cloud to respond to API requests before timeout.
    * **New Service** - Select:
        * **Percent** to specify a percentage of the target instances that you set in the **Elastigroup Setup** step.

>         * **Count** and specify an exact number of instances. This cannot exceed the **Max Instances** that you set in the **Elastigroup Setup** step.

Reviewer: I'm seeing a mix of dashes and full sentences when listing options. I typically use colons for all of them, but whatever you use, you should probably be consistent. Like this: 

- **Percent**: specify a percentage of the target instances that you set in the **Elastigroup Setup** step.

## Requirements and limitations

* Currently, Harness integrates with Spot only for Amazon Web Services (AWS) deployments by using Elastigroups. 
* Spot Elastigroup deployments support AMI artifacts only.
* Spot Elastigroup deployments support only Basic, Canary, and Blue Green execution types.
* Your spot account must be connected to the AWS cloud provider. For more information, go to [Connect your cloud account to Spot](https://docs.spot.io/connect-your-cloud-provider/aws-account).

## Connect to a Spot cloud provider

You can connect Harness to a Spot cloud provider by adding a [Harness Spot Elastigroup connector](/docs/platform/Connectors/Cloud-providers/add-a-spot-connector) or [AWS connector](/docs/platform/Connectors/Cloud-providers/add-aws-connector). 

To connect to a Spot cloud provider using Spot's API, go to [Spot API authentication](https://docs.spot.io/api/#section/Authentication).

To use an AWS connector, your spot account must first be connected to the AWS cloud provider. For more information, go to [Connect your cloud account to Spot](https://docs.spot.io/connect-your-cloud-provider/aws-account).

## Create the Harness Spot Elastigroup pipeline 

> You can create a service and environment when you are building the pipeline or separately in **Services** and **Environments**. In this topic, we walk through building these within a pipeline. 

Reviewer 2: ```suggestion
You can create a service and environment either when you are building the pipeline or separately in **Services** and **Environments**. This topic walks you through building these within a pipeline. 
```

The pipeline models the release process using execution steps, triggers, and other settings. For more information, go to [CD pipeline modeling overview](/docs/continuous-delivery/get-started/cd-pipeline-modeling-overview).

```mdx-code-block
import Tabs3 from '@theme/Tabs';
import TabItem3 from '@theme/TabItem';
```
```mdx-code-block
<Tabs3>
<TabItem3 value="Pipeline Studio" label="Pipeline Studio">
```

1. In your Harness Project, select **Deployments**.
2. Select **Pipelines**, and then select **Create a Pipeline**.
3. Enter the name **Spot Tutorial** for the pipeline, and then select **Start**.
4. Select **Add Stage**.
5. Select **Deploy**.
6. In **Stage Name**, enter **Spot Tutorial**.
7. In **Deployment Type**, select **Spot Elastigroup**, and then select **Set Up Stage**.

> The new stage is created. Next, we'll add a Harness service to represent the application you're deploying, and configure the service with the Spot Elastigroup service definition, artifacts, environment definition, and infrastructure for deployment.

Reviewer 2: ```suggestion
The new stage is created. Next, a Harness service is added to represent the application you're deploying and configure the service with the Spot Elastigroup service definition, artifacts, environment definition, and infrastructure for deployment.
```

Author: I want to keep the tone of a tutorial here, and also avid passive.

11. In **Test Region**, select an AWS region to test the credentials for this connector. 
12. In **Delegates Setup**, select **Only use Delegate with all of the following tags** and enter the name of the delegate created when creating the [Harness Spot Elastigroup connector](/docs/platform/Connectors/Cloud-providers/add-a-spot-connector).
13. Select **Save and Continue**
14.  After the connection test succeeds, select **Continue**.
15. In **Artifact Details**, enter the following details:
    1.  In **Artifact Source Identifier**, enter a name for the artifact.
    2.  In **Region**, select the region where the artifact is located. 
    3.  (Optional) In **AMI Tags**, add the tags used by your AMI. 
    4.  (Optional) In **AMI Filters**, add the AMI ID filters to locate the AMI resource.
    5.  In **Version Details**, select **Value** to enter the version number of the package you want to use. You can also select **Regex**, and then enter a tag regex to filter the package version.
    
    ![](./static/spot-artifacts-ami.png)
16. Select **Submit**.

> 17. **Save** the service configuration.

Reviewer 2: ```suggestion
17. Save the service configuration.
```

Reviewer 2: I think if we are referring to Save as the UI term, it should be "Select *Save* to save the configuration.

> 17. **Save** the service configuration.

Author: ```suggestion
17. Select **Save** to save the service configuration.
```

* Elastigroup Deploy: Specify how many instances to deploy, as a number or percentage of the Elastigroup parameters you've set up.

1. In Execution Strategies, select **Basic**, then select **Use Strategy**.
2. The basic execution steps are added. 
   
   ![](./static/spot-basic.png)

3. Select the **Elastigroup Setup** step to define **Step Parameters**.
    * **Name**: Enter the deployment step name.
    * **Timeout**: Set how long you want the Harness delegate to wait for the Spot cloud to respond to API requests before timeout.
    * **App Name**: Enter a name for the Elastigroup app name that Harness will create. For example, you can use Harness expressions, such as `<+project.identifier>_<+service.identifier>_<+env.identifier>_basic2`. You can select a fixed value, runtime input, or expression for this field. 
    * In **Instances**, select: 
       * **Same as already running instances**: Replicate the already running instances.
       * **Fixed**: Enter **Min Instances**, **Max Instances**, and **Desired Instances**.
4. Select the **Elastigroup Deploy** step to define **Step Parameters**.
    * **Name**: Enter the deployment step name.

>     * **Timeout**: Set how long you want the Harness delegate to wait for the Spot cloud to respond to API requests before timeout.

Reviewer 2: ```suggestion
    * **Timeout**: Set the timeout duration for the Harness delegate to wait for a response from the Spot cloud when making API requests.
```

* Elastigroup Deploy: Specify how many instances to deploy, as a number or percentage of the Elastigroup parameters you've set up.

1. In **Execution Strategies**, select **Canary**, and then click **Use Strategy**.
2. The canary execution steps are added. 
   
   ![](./static/spot-canary.png)
3. Select the **Elastigroup Setup** step to define **Step Parameters**.
    * **Name**: Enter the deployment step name.
    * **Timeout**: Set how long you want the Harness delegate to wait for the Spot cloud to respond to API requests before timeout.
    * **App Name**: Enter a name for the Elastigroup app name that Harness will create. For example, `<+project.identifier>_<+service.identifier>_<+env.identifier>_basic2`. You can select a fixed value, runtime input, or expression for this field. 
    * In **Instances**, select: 
       * **Same as already running instances**: Replicate the already running instances.
       * **Fixed**: Enter **Min Instances**, **Max Instances**, and **Desired Instances**.
4. Select the **Elastigroup Canary Deploy** step to define **Step Parameters**.
    * **Name**: Enter the deployment step name.

>     * **Timeout**: Set how long you want the Harness delegate to wait for the Spot cloud to respond to API requests before timeout.

Reviewer 2: ```suggestion
    * **Timeout**: Set the timeout duration for the Harness delegate to wait for a response from the Spot cloud when making API requests.
```

    * **New Service**: Select:
        * **Percent**: Specify a percentage of the target instances that you set in the **Elastigroup Setup** step.
        * **Count**: Specify the exact number of instances. This cannot exceed the **Max Instances** that you set in the **Elastigroup Setup** step.
5. Select the **Elastigroup Deploy** step to define **Step Parameters**.
    * **Name**: Enter the deployment step name.

>     * **Timeout**: Set how long you want the Harness delegate to wait for the Spot cloud to respond to API requests before timeout.

Reviewer 2: ```suggestion
    * **Timeout**: Set the timeout duration for the Harness delegate to wait for a response from the Spot cloud when making API requests.
```

Reviewer 2: If you are accepting this suggestion, please use it for all Timeout fields.
